deploy/eck-stack/charts/eck-agent/templates/elastic-agent.yaml

```diff
@@ -11,8 +11,84 @@ metadata:
     {{- toYaml . | nindent 4 }}
     {{- end }}
 spec:
-  version: {{ required "An Elastic Agent version is required" .Values.version }}
-  {{- if and (not (hasKey .Values.spec "daemonSet")) (not (hasKey .Values.spec "deployment")) (not (hasKey .Values.spec "statefulSet")) }}
-  {{ fail "At least one of daemonSet or deployment or statefulSet is required" }}
+  version: {{ required "An Elastic Agent version is required" (or ((.Values.spec).version) (.Values.version)) }}
+  {{- $daemonSet := (or (hasKey (.Values.spec) "daemonSet") (hasKey .Values "daemonSet")) }}
+  {{- $deployment := (or (hasKey (.Values.spec) "deployment") (hasKey .Values "deployment")) }}
+  {{- $statefulSet := (or (hasKey (.Values.spec) "statefulSet") (hasKey .Values "statefulSet")) }}
+  {{- if and (not $daemonSet) (not $deployment) (not $statefulSet) }}
+  {{ fail "At least one of daemonSet, deployment or statefulSet is required" }}
+  {{- end }}
+  {{- with or ((.Values.spec).count) (.Values.count) }}
+  count: {{ . }}
+  {{- end }}
+  {{- if $daemonSet }}
+  {{- $ds := or ((.Values.spec).daemonSet) (.Values.daemonSet) }}
+  daemonSet:
+    {{- /* This is required to render the empty daemonset ( {} ) properly */}}
+    {{- $ds | default dict | toYaml | nindent 4 }}
```

Comment:
This is what you get when removing that `default dict` stanza @thbkrkr .
```
                        Diff:
                                --- Expected
                                +++ Actual
                                @@ -4,3 +4,3 @@
                                       enabled: true
                                -daemonSet: {}
                                +daemonSet: null
```

This is what I mean by the comment, to get helm to render `{}` instead of `null`, you have to use this `default dict` syntax.

Reply:
I was testing with `daemonSet: {}` in `values.yaml`, but I see the issue if I just use `daemonSet:` with nothing.
I'm a bit torn between failing because `daemonSet:` sounds like a human mistake to surface or keeping `default dict` as a safety measure. Not strong opinion on this.

Reply:
Let me test and see if there's a better way to handle this. I spent some time on this as it wasn't a simple thing to handle in helm...

Reply:
I agree with using `default dict` as a safety measure to handle the case where daemonSet is set to null.